keps/sig-node/127-usernamespaces-support/README.md

```diff
@@ -0,0 +1,1017 @@
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+
+This section only focuses on phase 1 as specified above.
+
+### Summary of the Proposed Changes
+
+- Extend the CRI to have a user namespace mode and the user and group ID mappings.
+- Add a `userNamespaceMode` field to the pod spec.
+- Add the cluster-wide ID mappings to the kubelet configuration file.
+- Add a `UserNamespacesSupport` feature flag to enable / disable the user
+  namespaces support.
+- Update owner of ephemeral volumes populated by the kubelet.
+
+### CRI API Changes
+
+The CRI is extended to (optionally) specify the user namespace mode
+and the ID mappings for a pod.
+[`NamespaceOption`](https://github.com/kubernetes/cri-api/blob/1eae59a7c4dee45a900f54ea2502edff7e57fd68/pkg/apis/runtime/v1alpha2/api.proto#L228)
+is extended with two new fields:
+- A `user` `NamespaceMode` that defines if the pod should run in an independent
+  user namespace (`POD`) or if it should share the host user namespace
+  (`NODE`).
+- The ID mappings to be used if the user namespace mode is `POD`.
+
+```
+// LinuxIDMapping represents a single user namespace mapping in Linux.
+message LinuxIDMapping {
+   // container_id is the starting ID for the mapping inside the container.
+   uint32 container_id = 1;
+   // host_id is the starting ID for the mapping on the host.
+   uint32 host_id = 2;
+   // number of IDs in this mapping.
+   uint32 length = 3;
+}
+
+// LinuxUserNamespaceConfig represents the user and group ID mapping in Linux.
+message LinuxUserNamespaceConfig {
+   // uid_mappings is an array of user ID mappings.
+   repeated LinuxIDMapping uid_mappings = 1;
+   // gid_mappings is an array of group ID mappings.
+   repeated LinuxIDMapping gid_mappings = 2;
+}
+
+// NamespaceOption provides options for Linux namespaces.
+message NamespaceOption {
+    ...
+    // User namespace for this container/sandbox.
+    // Note: There is currently no way to set CONTAINER scoped user namespace in the Kubernetes API.
+    // Namespaces currently set by the kubelet: POD, NODE
+    Namespace user = 5;
+    // ID mappings to use when the user namespace mode is POD.
+    LinuxUserNamespaceConfig mappings = 6;
+}
+```
+
+### PodSpec Changes
+
+`v1.PodSpec` is extended with a new `UserNamespaceMode` field:
+
+```
+const (
+	UserNamespaceModeHost    PodUserNamespaceMode = "Host"
+	UserNamespaceModeCluster PodUserNamespaceMode = "Cluster"
+)
+
+type PodSpec struct {
+...
+  // UserNamespaceMode controls how user namespaces are used for this Pod.
+  // Three modes are supported:
+  // "Host": The pod shares the host user namespace. (default value).
+  // "Cluster": The pod uses a cluster-wide configured ID mappings.
+  // +k8s:conversion-gen=false
+  // +optional
+  UserNamespaceMode PodUserNamespaceMode `json:"userNamespaceMode,omitempty" protobuf:"bytes,36,opt  name=userNamespaceMode"`
```

Comment:
This should be under the security context for the pod.

Reply:
I put it there because other similar fields like `host{Network, IPC, PID}` are at the `PodSpec` level. I don't have a strong opinion but I find it a bit "weird" to control some namespaces at one level and the others in a different level.